src/ConstantInterval.cpp

```diff
@@ -0,0 +1,643 @@
+    result.max = 0;
+    return result;
+}
+
+ConstantInterval ConstantInterval::bounded_above(int64_t max) {
+    ConstantInterval result(max, max);
+    result.min_defined = false;
+    result.min = 0;
+    return result;
+}
+
+bool ConstantInterval::is_everything() const {
+    return !min_defined && !max_defined;
+}
+
+bool ConstantInterval::is_single_point() const {
+    return min_defined && max_defined && min == max;
+}
+
+bool ConstantInterval::is_single_point(int64_t x) const {
+    return min_defined && max_defined && min == x && max == x;
+}
+
+bool ConstantInterval::is_bounded() const {
+    return max_defined && min_defined;
+}
+
+bool ConstantInterval::operator==(const ConstantInterval &other) const {
+    if (min_defined != other.min_defined || max_defined != other.max_defined) {
+        return false;
+    }
+    return (!min_defined || min == other.min) && (!max_defined || max == other.max);
+}
+
+void ConstantInterval::include(const ConstantInterval &i) {
+    if (max_defined && i.max_defined) {
+        max = std::max(max, i.max);
+    } else {
+        max_defined = false;
+    }
+    if (min_defined && i.min_defined) {
+        min = std::min(min, i.min);
+    } else {
+        min_defined = false;
+    }
+}
+
+void ConstantInterval::include(int64_t x) {
+    if (max_defined) {
+        max = std::max(max, x);
+    }
+    if (min_defined) {
+        min = std::min(min, x);
+    }
+}
+
+bool ConstantInterval::contains(int64_t x) const {
+    return !((min_defined && x < min) ||
+             (max_defined && x > max));
+}
+
+bool ConstantInterval::contains(int32_t x) const {
+    return contains((int64_t)x);
+}
+
+bool ConstantInterval::contains(uint64_t x) const {
+    if (x <= (uint64_t)std::numeric_limits<int64_t>::max()) {
+        return contains((int64_t)x);
+    } else {
+        return !max_defined;
+    }
+}
+
+ConstantInterval ConstantInterval::make_union(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result = a;
+    result.include(b);
+    return result;
+}
+
+ConstantInterval ConstantInterval::make_intersection(const ConstantInterval &a,
+                                                     const ConstantInterval &b) {
+    ConstantInterval result;
+    if (a.min_defined) {
+        if (b.min_defined) {
+            result.min = std::max(a.min, b.min);
+        } else {
+            result.min = a.min;
+        }
+        result.min_defined = true;
+    } else {
+        result.min_defined = b.min_defined;
+        result.min = b.min;
+    }
+    if (a.max_defined) {
+        if (b.max_defined) {
+            result.max = std::min(a.max, b.max);
+        } else {
+            result.max = a.max;
+        }
+        result.max_defined = true;
+    } else {
+        result.max_defined = b.max_defined;
+        result.max = b.max;
+    }
+    internal_assert(!result.is_bounded() || result.min <= result.max)
+        << "Empty ConstantInterval constructed in make_intersection";
```

Comment:
Is this error message right? The first branch of the `||` is true if the result is unbounded, i.e. empty?

Reply:
unbounded means goes to infinity on at least one side, not empty. I'm not sure how we'd represent empty, but I think it would be min_defined && max_defined && min > max.